### PR TITLE
fix(train): delete merge-queue branch with no car attached

### PR DESCRIPTION
**Pull request by Syffe**
A bug existed that didn't persistently deleted merge-queue branches
that weren't attached to any car/train after run.

Fixes MRGFY-938

### CODE REVIEW

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 1]
```diff
@@ -3500,6 +3500,93 @@ async def test_create_pull_basic(self):
...
+
+        queue_rule = rules.QueueRule(
+            name="foo",
+            conditions=conditions.QueueRuleConditions([]),
+            config=queue_config,
+        )
+        await car1.create_pull(queue_rule)
+        await car2.create_pull(queue_rule)
+
+        from urllib import parse
```
Comment (sileht) [inline mergify_engine/tests/functional/actions/test_queue.py:3559]: Import should go on top

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 2]
```diff
@@ -3500,6 +3500,93 @@ async def test_create_pull_basic(self):
...
+
+        match = (
+            f"{constants.MERGE_QUEUE_BRANCH_PREFIX}/" f"{parse.quote(q.ref, safe='')}"
+        )
+
```
Comment (sileht) [inline mergify_engine/tests/functional/actions/test_queue.py:3564]: I think you should also create other branches that your code will not delete.

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 3]
```diff
@@ -3500,6 +3500,93 @@ async def test_create_pull_basic(self):
...
+            branch
+            async for branch in typing.cast(
+                typing.AsyncGenerator[github_types.GitHubGitRef, None],
+                self.repository_ctxt.installation.client.items(
+                    f"/repos/{self.repository_ctxt.installation.owner_login}/{self.repository_ctxt.repo['name']}/git/matching-refs/heads/{match}",
+                ),
+            )
+        ]
+
+        assert len(merge_queue_branch_list) == 2
```
Comment (sileht) [inline mergify_engine/tests/functional/actions/test_queue.py:3575]: You can use `self.get_branches()` to list all branches and assert the exact expected list. (eg: "main", `self.main.branch_name`, the branches of the draft PRs, the other unrelated branches you delete).

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 4]
```diff
@@ -3500,6 +3500,93 @@ async def test_create_pull_basic(self):
...
+        merge_queue_branch_list = [
+            branch
+            async for branch in typing.cast(
+                typing.AsyncGenerator[github_types.GitHubGitRef, None],
+                self.repository_ctxt.installation.client.items(
+                    f"/repos/{self.repository_ctxt.installation.owner_login}/{self.repository_ctxt.repo['name']}/git/matching-refs/heads/{match}",
+                ),
+            )
+        ]
+        assert merge_queue_branch_list == []
```
Comment (sileht) [inline mergify_engine/tests/functional/actions/test_queue.py:3588]: You should assert the exact list of all branches remaining in the repository, like I said before.

**Review by sileht — DISMISSED** [review]

File: mergify_engine/tests/functional/actions/test_queue.py [hunk 5]
```diff
@@ -3591,6 +3591,103 @@ async def test_create_pull_basic(self):
...
+        )
+
+        await car1.create_pull(queue_rule)
+        await car2.create_pull(queue_rule)
+        await car3.create_pull(queue_rule)
+        await car4.create_pull(queue_rule)
+
+        assert 6 == len(await self.get_branches())
+        assert 3 == len(q._cars)
+
```
Comment (sileht) [inline mergify_engine/tests/functional/actions/test_queue.py:3685]: You also need to create branches unrelated to the main branch and to the train cars.

(setup_repo() has an attribute called test_branches that you can leverage for this.)

File: mergify_engine/tests/unit/test_train.py [hunk 6]
```diff
@@ -193,6 +193,7 @@ def update_base_sha(sha):
 
     client = mock.Mock()
     client.item = mock.AsyncMock(side_effect=item_call)
+    client.items = mock.MagicMock()
```
Comment (sileht) [inline mergify_engine/tests/unit/test_train.py:196]: 👍

**Review by sileht — CHANGES_REQUESTED** [review]

File: mergify_engine/tests/functional/actions/test_queue.py (hunk 5 above)
Comment (Syffe) [reply mergify_engine/tests/functional/actions/test_queue.py:3685]: Indeed, forgot to create those ! Gonna change this !

**Comment by sileht** [comment]
I just added a test with an empty train.

**Review by sileht — APPROVED** [review]

**Pull request merged**